File: plugins/splash/plugin/ios/DriftSplashPlugin.swift
/// DriftSplashPlugin.swift
///
/// Entry point invoked from the generated `DriftPluginRegistrant.registerAll`.
/// Responsibilities:
///   1. Install the overlay UIView synchronously on the active window before
///      the system launch screen tears down. (Eager
///      `PlatformChannelManager.shared` access in SceneDelegate guarantees
///      this register call runs in time.)
///   2. Register the `drift/splash` channel handler. The Go runtime calls
///      `preserve` / `remove`; both forward to `SplashState.apply(±1)`.
///   3. Subscribe to `drift/rendering/frame_events`. On `first_frame`, mark
///      state and dismiss if there are no outstanding preserves.

import OSLog
import UIKit

private let splashLog = OSLog(subsystem: "drift.splash", category: "plugin")

enum DriftSplashPlugin {

    private static var overlay: DriftSplashOverlayView?
    private static var frameSubscription: DriftSubscription?

    static func register(host: DriftPluginHost) {
        installOverlay(host: host)
        registerChannelHandler(host: host)
        observeFirstFrame(host: host)
    }

    /// 1. Install the overlay synchronously. DriftOverlayHost is an optional
    ///    capability protocol; bail with a logged warning if the host
    ///    doesn't adopt it or has no active root view. Channel + observer
    ///    still install in those cases so Preserve/Remove + state tracking
    ///    keep working; only the visible overlay is missing.
    private static func installOverlay(host: DriftPluginHost) {
        guard let overlayHost = host as? DriftOverlayHost else {
            os_log("host does not implement DriftOverlayHost; runtime overlay disabled (likely framework/plugin version mismatch)",
                   log: splashLog, type: .info)
            return
        }
        guard let rootView = overlayHost.driftRootView() else {
            os_log("DriftOverlayHost.driftRootView() returned nil; runtime overlay disabled (no active scene?)",
                   log: splashLog, type: .info)
            return
        }
        let view = DriftSplashOverlayView()
        view.frame = rootView.bounds
        view.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        rootView.addSubview(view)
        overlay = view
    }

    /// 2. Channel handler: named methods matching the Go API.
    ///    SplashState.apply(±1) is the single saturation site; clamps a
    ///    Remove without a matching Preserve to zero.
    private static func registerChannelHandler(host: DriftPluginHost) {
        host.registerChannel("drift/splash") { method, _ in
            switch method {
            case "preserve":
                DriftSplashState.shared.apply(1)
                maybeDismiss()
                return (nil, nil)
            case "remove":
                DriftSplashState.shared.apply(-1)
                maybeDismiss()
                return (nil, nil)
            default:
                return (nil, NSError(domain: "drift.splash", code: 1, userInfo: [
                    NSLocalizedDescriptionKey: "unknown splash method \(method)",
                ]))
            }
        }
    }

    /// 3. Subscribe to frame events via the host. The subscription token is
    ///    retained on the plugin so cleanup after iOS scene-recreate
    ///    scenarios in v2 stays straightforward.
    private static func observeFirstFrame(host: DriftPluginHost) {
        frameSubscription = host.observeEvent("drift/rendering/frame_events") { data in
            guard let payload = data as? [String: Any],
                  let type = payload["type"] as? String,
                  type == "first_frame" else { return }
            DriftSplashState.shared.markFirstFrame()
            maybeDismiss()
        }
    }

    private static func maybeDismiss() {
        guard DriftSplashState.shared.canDismiss() else { return }
        DispatchQueue.main.async {
            guard UIApplication.shared.applicationState == .active else { return }
            overlay?.fadeOut(durationMs: DriftSplashConfig.fadeDurationMs) {
                overlay = nil
            }
        }
    }
}

File: plugins/splash/plugin/android/DriftSplashPlugin.kt
/**
 * DriftSplashPlugin.kt
 *
 * Entry point invoked from the generated DriftPluginRegistrant.registerAll.
 * Responsibilities (mirrors the iOS implementation):
 *   1. Install the overlay View on `host.driftRootView()` (the activity's
 *      decorView) synchronously so the visual hand-off from the legacy
 *      LaunchTheme drawable is seamless.
 *   2. Register the `drift/splash` channel handler. The Go runtime calls
 *      `preserve` / `remove`; both forward to `DriftSplashState.apply(±1)`.
 *   3. Subscribe natively to `drift/rendering/frame_events`. On `first_frame`,
 *      mark state and dismiss if there are no outstanding preserves.
 */
package com.drift.plugin.splash

import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.ViewGroup
import com.drift.runner.DriftOverlayHost
import com.drift.runner.DriftPluginHost
import com.drift.runner.DriftSubscription

object DriftSplashPlugin {

    private const val TAG = "DriftSplash"

    private var overlay: DriftSplashOverlayView? = null
    private var frameSubscription: DriftSubscription? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    @JvmStatic
    fun register(host: DriftPluginHost) {
        installOverlay(host)
        // Channel handler: named methods matching the Go API.
        // DriftSplashState.apply(±1) is the single saturation site;
        // a Remove without a matching Preserve clamps to zero.
        host.registerChannel("drift/splash") { method, _ ->
            when (method) {
                "preserve" -> {
                    DriftSplashState.apply(1)
                    maybeDismiss()
                    Pair(null, null)
                }
                "remove" -> {
                    DriftSplashState.apply(-1)
                    maybeDismiss()
                    Pair(null, null)
                }
                else -> Pair(null, IllegalArgumentException("unknown splash method $method"))
            }
        }
        // SkiaHostView fires first_frame after the next frame commit; the
        // host's event router delivers it here. The subscription token is
        // retained so v2 cleanup (Activity-recreate teardown) has a handle.
        frameSubscription = host.observeEvent("drift/rendering/frame_events") { data ->
            val payload = data as? Map<*, *> ?: return@observeEvent
            if (payload["type"] != "first_frame") return@observeEvent
            DriftSplashState.markFirstFrame()
            maybeDismiss()
        }
    }

    private fun installOverlay(host: DriftPluginHost) {
        val overlayHost = host as? DriftOverlayHost
        if (overlayHost == null) {
            Log.w(TAG, "host does not implement DriftOverlayHost; runtime overlay disabled " +
                "(likely framework/plugin version mismatch)")
            return
        }
        val rootView = overlayHost.driftRootView() as? ViewGroup
        if (rootView == null) {
            Log.w(TAG, "DriftOverlayHost.driftRootView() returned null or non-ViewGroup; " +
                "runtime overlay disabled (no active activity?)")
            return
        }
        val view = DriftSplashOverlayView(rootView.context)
        rootView.addView(view)
        overlay = view
    }

    private fun maybeDismiss() {
        if (!DriftSplashState.canDismiss()) return
        mainHandler.post {
            val current = overlay ?: return@post
            current.fadeOut(DriftSplashConfig.FADE_DURATION_MS) { overlay = null }
        }
    }
}
